fix: refactor top tv shows picks

Abdulrahman-Ragab-01 opened the pull request:
# What does this PR do?
<!-- Brief description of changes -->
refactor top tv shows picks presentation

## Type of Change
- [x] 🔧 Refactoring

## Testing
- [x] Unit tests added/updated
- [x] UI tests added/updated

## Checklist
- [x] Tests pass

Abdulrahman-Ragab-01 committed:
fix: refactor top tv shows picks

Abdulrahman-Ragab-01 committed:
fix: improve test coverage

Abdulrahman-Ragab-01 committed:
fix: test cases conflicts

ahmeddsamii commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksContract.kt:
```diff
@@ -1,8 +1,8 @@
-    fun onSaveTvShow(tvShowId: Int)
-    fun onTvShowClicked(tvShowId: Int)
-    fun onBack()
     fun onRetry()
```
rename all function to have the same suffix

ahmeddsamii commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksEffect.kt:
```diff
@@ -1,6 +1,6 @@
-}
\ No newline at end of file
+    data object NavigateBack : TopTvShowsPicksEffect
+    data class NavigateToTvShowDetails(val tvShowId: Int) : TopTvShowsPicksEffect
```
the class name should be noun not verb

Abdulrahman-Ragab-01 committed:
fix: clean code and increase test coverage

Abdulrahman-Ragab-01 committed:
fix: rename content function

AsmaaKhaledGomaa commented on presentation/src/test/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksViewModelTest.kt:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    @Test
+    fun `when viewModel is initialized with null args should use default actorId`() = runTest {
```
Please rename the test function to
function name should return ... when onSaveClick ...

Abdulrahman-Ragab-01 committed:
fix: rename test cases

Abdulrahman-Ragab-01 committed:
fix: delete useless modifier

Abdulrahman-Ragab-01 committed:
fix: increase test coverage

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

marwanqashwa commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksScreen.kt:
```diff
@@ -15,59 +14,57 @@ import com.london.presentation.utils.Listen
 fun TopTvShowsPicksScreen(
-    onNavigateTvShow: (Int) -> Unit,
     onNavigateBack: () -> Unit,
+    onNavigateTvShow: (Int) -> Unit,
```
i think onNavigateTvShowDetails

marwanqashwa commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksScreen.kt:
```diff
@@ -15,59 +14,57 @@ import com.london.presentation.utils.Listen
             title = stringResource(R.string.top_tv_shows_picks),
-            items = state.tvShowDetails.cast,
-            onBack = contract::onBack,
+            items = state.actorTvShowDetails.cast,
```
same i don't understand what is the cast i think screen have only tv shows

Abdulrahman-Ragab-01 replied:
Done, I solve this in actor details branch

Abdulrahman-Ragab-01 replied:
check now, its solved in this branch too

marwanqashwa commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksUiState.kt:
```diff
@@ -4,10 +4,8 @@ import com.london.domain.entity.actordetails.cast.CastDetails
-    val tvShowDetails: CastDetails = CastDetails(),
-    val errorState: ErrorState? = null,
-    val isLoading: Boolean = false,
     val isSaved: Boolean = false,
```
this save for all tv shows in screen

marwanqashwa commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksViewModel.kt:
```diff
@@ -25,39 +25,28 @@ class TopTvShowsPicksViewModel @Inject constructor(
     }
 
-    override fun onSaveTvShow(tvShowId: Int) {
+    override fun onSaveClick(tvShowId: Int) {
```
i mean this

Abdulrahman-Ragab-01 committed:
fix: rename functions and delete unused issaved variable in uistate and viewmodel

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh commented on app/src/main/java/com/london/app/navigation/NovixApp.kt:
```diff
@@ -439,7 +439,7 @@ fun NavGraphBuilder.mainNavGraph(
         popExitTransition = { fadeOut(tween(500)) },
     ) {
         TopTvShowsPicksScreen(
```
if this screen not used multiple times, move the navigation in to the screen

Mostafa-alsaygh commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksEffect.kt:
```diff
@@ -1,6 +1,6 @@
     data object BackNavigation : TopTvShowsPicksEffect
-}
\ No newline at end of file
+    data class TvShowDetailsNavigation(val tvShowId: Int) : TopTvShowsPicksEffect
```
the name is not right see the record in 31:00 min of bareq review

Mostafa-alsaygh commented on presentation/src/test/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksViewModelTest.kt:
```diff
@@ -0,0 +1,106 @@
+    private val testScheduler = TestCoroutineScheduler()
+    private val testDispatcher = UnconfinedTestDispatcher(testScheduler)
+
+    private val mockCastDetails = CastDetails(
```
move the fake data in the last of file of in separate one

Abdulrahman-Ragab-01 replied:
done

Mostafa-alsaygh commented on presentation/src/test/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksViewModelTest.kt:
```diff
@@ -0,0 +1,106 @@
+        viewModel = TopTvShowsPicksViewModel(savedStateHandle, getActorTvShowPicksById)
+
+        // Then
+        viewModel.onRetryClick()
```
thats not right see my PR 
u should Listen to the effect

Mostafa-alsaygh commented on presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksScreen.kt:
```diff
@@ -15,59 +14,57 @@ import com.london.presentation.utils.Listen
     val state by viewModel.state.collectAsStateWithLifecycle()
     val effect by viewModel.effect.collectAsState(initial = null)
 
     HandleTvShowsPicksEffects(
```
remove this fun and make directly

Abdulrahman-Ragab-01 replied:
done

Mostafa-alsaygh commented:
why the prefix name is fix and the name is refactor

harrytmthy commented on presentation/src/test/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksViewModelTest.kt:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Test
+    fun `should show loading state when data is being fetched`() = runTest {
```
This test has no assertions. If it's meant to verify the "loading" state, consider comparing the state before and after the fetch.

harrytmthy commented on presentation/src/test/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksViewModelTest.kt:
```diff
@@ -0,0 +1,106 @@
+        val args = Screen.TopTvShowsPicksDetails(actorId)
+        every { savedStateHandle.getArgs<Screen.TopTvShowsPicksDetails>() } returns args
+        coEvery { getActorTvShowPicksById.invoke(actorId) } coAnswers {
+            delay(100)
```
Rather than just using a real `delay(100)` in the stub, consider simulating the delay with virtual time using `advanceTimeBy(100)` or `advanceUntilIdle()`. This keeps the test fast and deterministic.

Abdulrahman-Ragab-01 replied:
thank you for this info

Abdulrahman-Ragab-01 committed:
fix: refactor test cases

Mostafa-alsaygh committed:
fix/tv-shows-picks-testing (#631)

refactor: improve unit tests for TopTvShowsPicksViewModel

This commit refactors the unit tests for `TopTvShowsPicksViewModel` by:

-   Replacing `TestCoroutineScheduler` and `UnconfinedTestDispatcher` with `StandardTestDispatcher` for better control over coroutine execution.
-   Utilizing `advanceUntilIdle()` to ensure all coroutines complete before making assertions.
-   Adding tests for error handling scenarios.
-   Improving test structure and readability.

Abdulrahman-Ragab-01 committed:
Merge remote-tracking branch 'origin/develop' into hotfix/top-tv-shows-picks

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/details/actor/info/toptvshowspicks/TopTvShowsPicksUiState.kt

Abdulrahman-Ragab-01 committed:
fix: rearrange override functions in viewmodel and handle delete effect
